File: src/desktop/modern/src/core/dependency_injection/service_registry.py
"""
Service Registry - Centralized Service Registration Management

Handles all service registration operations including:
- Singleton registration
- Transient registration
- Factory registration
- Instance registration
- Scoped registration
- Lazy registration
"""

from typing import Type, TypeVar, Dict, Any, Optional, Callable
from enum import Enum
from dataclasses import dataclass
import logging

logger = logging.getLogger(__name__)

T = TypeVar("T")


class ServiceScope(Enum):
    """Service scope definitions for advanced DI features."""

    SINGLETON = "singleton"
    TRANSIENT = "transient"
    REQUEST = "request"
    SESSION = "session"


@dataclass
class ServiceDescriptor:
    """Enhanced service descriptor with scope and caching support."""

    interface: Type
    implementation: Type
    scope: ServiceScope = ServiceScope.SINGLETON
    factory: Optional[Callable] = None
    lazy: bool = False


class ServiceRegistry:
    """
    Centralized service registration management.

    Handles all types of service registrations and maintains the service mappings
    used by the DI container for resolution.
    """

    def __init__(self):
        # Core service storage
        self.services: Dict[Type, Type] = {}  # Singleton services
        self.singletons: Dict[Type, Any] = {}  # Singleton instances
        self._factories: Dict[Type, Type] = {}  # Transient/factory services

        # Advanced features
        self._service_descriptors: Dict[Type, ServiceDescriptor] = {}
        self._scoped_instances: Dict[str, Dict[Type, Any]] = {}
        self._current_scope: Optional[str] = None

    def register_singleton(self, interface: Type[T], implementation: Type[T]) -> None:
        """Register a service as singleton (one instance per container)."""
        self.services[interface] = implementation
        logger.debug(
            f"Registered singleton: {interface.__name__} -> {implementation.__name__}"
        )

    def register_transient(self, interface: Type[T], implementation: Type[T]) -> None:
        """Register a service as transient (new instance each time)."""
        self._factories[interface] = implementation
        logger.debug(
            f"Registered transient: {interface.__name__} -> {implementation.__name__}"
        )

    def register_instance(self, interface: Type[T], instance: T) -> None:
        """Register a specific instance."""
        self.singletons[interface] = instance
        logger.debug(f"Registered instance: {interface.__name__}")

    def register_factory(self, interface: Type[T], factory_func: Callable) -> None:
        """Register a factory function for creating instances."""
        self._factories[interface] = factory_func
        logger.debug(f"Registered factory: {interface.__name__}")

    def register_scoped(
        self, interface: Type[T], implementation: Type[T], scope: ServiceScope
    ) -> None:
        """Register a service with specific scope (singleton, transient, request, session)."""
        descriptor = ServiceDescriptor(
            interface=interface, implementation=implementation, scope=scope
        )
        self._service_descriptors[interface] = descriptor

        # Also register in appropriate legacy collection for compatibility
        if scope == ServiceScope.SINGLETON:
            self.services[interface] = implementation
        else:
            self._factories[interface] = implementation

        logger.debug(
            f"Registered {scope.value}: {interface.__name__} -> {implementation.__name__}"
        )

    def register_lazy(self, interface: Type[T], implementation: Type[T]) -> None:
        """Register a service for lazy loading."""
        descriptor = ServiceDescriptor(
            interface=interface,
            implementation=implementation,
            scope=ServiceScope.SINGLETON,
            lazy=True,
        )
        self._service_descriptors[interface] = descriptor
        logger.debug(
            f"Registered lazy: {interface.__name__} -> {implementation.__name__}"
        )

    def is_registered(self, interface: Type) -> bool:
        """Check if a service is registered."""
        return (
            interface in self.services
            or interface in self._factories
            or interface in self.singletons
        )

    def get_service_implementation(self, interface: Type) -> Optional[Type]:
        """Get the implementation type for a service interface."""
        if interface in self.services:
            return self.services[interface]
        elif interface in self._factories:
            return self._factories[interface]
        return None

    def get_singleton_instance(self, interface: Type) -> Optional[Any]:
        """Get existing singleton instance if available."""
        return self.singletons.get(interface)

    def set_singleton_instance(self, interface: Type, instance: Any) -> None:
        """Store a singleton instance."""
        self.singletons[interface] = instance

    def has_singleton_instance(self, interface: Type) -> bool:
        """Check if singleton instance exists."""
        return interface in self.singletons

    def has_factory_registration(self, interface: Type) -> bool:
        """Check if service has factory registration."""
        return interface in self._factories

    def has_service_registration(self, interface: Type) -> bool:
        """Check if service has singleton registration."""
        return interface in self.services

    def get_factory_or_implementation(self, interface: Type) -> Optional[Any]:
        """Get factory function or implementation class."""
        return self._factories.get(interface)

    def get_all_registrations(self) -> Dict[Type, Type]:
        """Get all registered services for debugging."""
        registrations = {}

        # Add singleton instances
        for interface in self.singletons.keys():
            registrations[interface] = type(self.singletons[interface])

        # Add service registrations
        registrations.update(self.services)

        # Add factory registrations
        registrations.update(self._factories)

        return registrations

    def get_service_descriptor(self, interface: Type) -> Optional[ServiceDescriptor]:
        """Get service descriptor for advanced features."""
        return self._service_descriptors.get(interface)

    def clear_all(self) -> None:
        """Clear all registrations (useful for testing)."""
        self.services.clear()
        self.singletons.clear()
        self._factories.clear()
        self._service_descriptors.clear()
        self._scoped_instances.clear()
        self._current_scope = None
        logger.debug("All service registrations cleared")

    def get_registration_count(self) -> int:
        """Get total number of registrations."""
        return len(self.services) + len(self._factories) + len(self.singletons)

File: src/desktop/shared/di/core/ServiceRegistry.ts
/**
 * 📋 TKA SERVICE REGISTRY
 *
 * Advanced service registration and metadata management system
 * that maintains comprehensive information about all registered services.
 */

import {
    ServiceInterface,
    ServiceDescriptor,
    ServiceScope,
    ServiceMetadata,
    Constructor,
    ServiceFactory
} from './types.js';

export class ServiceRegistry {
    private readonly _descriptors = new Map<string, ServiceDescriptor>();
    private readonly _singletonInstances = new Map<string, any>();
    private readonly _scopedInstances = new Map<string, Map<string, any>>();
    private readonly _registrationOrder: string[] = [];

    // Advanced indexing for fast lookups
    private readonly _servicesByScope = new Map<ServiceScope, Set<string>>();
    private readonly _servicesByTag = new Map<string, Set<string>>();
    private readonly _dependencyGraph = new Map<string, Set<string>>();

    // ============================================================================
    // CORE REGISTRATION METHODS
    // ============================================================================

    /**
     * Register a singleton service
     */
    registerSingleton<T>(
        serviceInterface: ServiceInterface<T>,
        implementation: Constructor<T>
    ): void {
        this._registerService(serviceInterface, implementation, ServiceScope.Singleton);
    }

    /**
     * Register a transient service
     */
    registerTransient<T>(
        serviceInterface: ServiceInterface<T>,
        implementation: Constructor<T>
    ): void {
        this._registerService(serviceInterface, implementation, ServiceScope.Transient);
    }

    /**
     * Register a scoped service
     */
    registerScoped<T>(
        serviceInterface: ServiceInterface<T>,
        implementation: Constructor<T>,
        scope: ServiceScope
    ): void {
        this._registerService(serviceInterface, implementation, scope);
    }

    /**
     * Register a factory function
     */
    registerFactory<T>(
        serviceInterface: ServiceInterface<T>,
        factory: ServiceFactory<T>,
        scope: ServiceScope = ServiceScope.Singleton
    ): void {
        const descriptor: ServiceDescriptor<T> = {
            serviceInterface,
            factory,
            scope,
            registeredAt: new Date(),
            metadata: serviceInterface.metadata
        };

        this._storeDescriptor(serviceInterface.name, descriptor);
        this._updateIndexes(serviceInterface.name, scope, serviceInterface.metadata);
    }

    /**
     * Register a specific instance
     */
    registerInstance<T>(
        serviceInterface: ServiceInterface<T>,
        instance: T
    ): void {
        const descriptor: ServiceDescriptor<T> = {
            serviceInterface,
            instance,
            scope: ServiceScope.Instance,
            registeredAt: new Date(),
            metadata: serviceInterface.metadata
        };

        this._storeDescriptor(serviceInterface.name, descriptor);
        this._singletonInstances.set(serviceInterface.name, instance);
        this._updateIndexes(serviceInterface.name, ServiceScope.Instance, serviceInterface.metadata);
    }

    /**
     * Register a lazy service
     */
    registerLazy<T>(
        serviceInterface: ServiceInterface<T>,
        implementation: Constructor<T>
    ): void {
        this._registerService(serviceInterface, implementation, ServiceScope.Lazy);
    }

    // ============================================================================
    // INSTANCE MANAGEMENT
    // ============================================================================

    /**
     * Get singleton instance if exists
     */
    getSingletonInstance<T>(serviceInterface: ServiceInterface<T>): T | null {
        return this._singletonInstances.get(serviceInterface.name) || null;
    }

    /**
     * Store singleton instance
     */
    setSingletonInstance<T>(serviceInterface: ServiceInterface<T>, instance: T): void {
        this._singletonInstances.set(serviceInterface.name, instance);
    }

    /**
     * Get scoped instance if exists
     */
    getScopedInstance<T>(serviceInterface: ServiceInterface<T>, scopeId: string): T | null {
        const scopeMap = this._scopedInstances.get(scopeId);
        return scopeMap?.get(serviceInterface.name) || null;
    }

    /**
     * Store scoped instance
     */
    setScopedInstance<T>(serviceInterface: ServiceInterface<T>, scopeId: string, instance: T): void {
        let scopeMap = this._scopedInstances.get(scopeId);
        if (!scopeMap) {
            scopeMap = new Map();
            this._scopedInstances.set(scopeId, scopeMap);
        }
        scopeMap.set(serviceInterface.name, instance);
    }

    /**
     * Clear scoped instances for a specific scope
     */
    clearScopedInstances(scopeId: string): void {
        this._scopedInstances.delete(scopeId);
    }

    // ============================================================================
    // QUERY AND LOOKUP METHODS
    // ============================================================================

    /**
     * Check if a service is registered
     */
    isRegistered<T>(serviceInterface: ServiceInterface<T>): boolean {
        return this._descriptors.has(serviceInterface.name);
    }

    /**
     * Get service descriptor
     */
    getDescriptor<T>(serviceInterface: ServiceInterface<T>): ServiceDescriptor<T> | null {
        return this._descriptors.get(serviceInterface.name) || null;
    }

    /**
     * Get all registered services
     */
    getRegisteredServices(): string[] {
        return Array.from(this._descriptors.keys());
    }

    /**
     * Get services by scope
     */
    getServicesByScope(scope: ServiceScope): string[] {
        return Array.from(this._servicesByScope.get(scope) || []);
    }

    /**
     * Get services by tag
     */
    getServicesByTag(tag: string): string[] {
        return Array.from(this._servicesByTag.get(tag) || []);
    }

    /**
     * Get service dependencies
     */
    getServiceDependencies(serviceName: string): string[] {
        return Array.from(this._dependencyGraph.get(serviceName) || []);
    }

    /**
     * Get registration order
     */
    getRegistrationOrder(): string[] {
        return [...this._registrationOrder];
    }

    // ============================================================================
    // ADVANCED QUERIES
    // ============================================================================

    /**
     * Find services matching criteria
     */
    findServices(criteria: {
        scope?: ServiceScope;
        tag?: string;
        deprecated?: boolean;
        registeredAfter?: Date;
        registeredBefore?: Date;
    }): ServiceDescriptor[] {
        const results: ServiceDescriptor[] = [];

        for (const descriptor of this._descriptors.values()) {
            let matches = true;

            if (criteria.scope && descriptor.scope !== criteria.scope) {
                matches = false;
            }

            if (criteria.tag && !descriptor.metadata?.tags?.includes(criteria.tag)) {
                matches = false;
            }

            if (criteria.deprecated !== undefined && descriptor.metadata?.deprecated !== criteria.deprecated) {
                matches = false;
            }

            if (criteria.registeredAfter && descriptor.registeredAt < criteria.registeredAfter) {
                matches = false;
            }

            if (criteria.registeredBefore && descriptor.registeredAt > criteria.registeredBefore) {
                matches = false;
            }

            if (matches) {
                results.push(descriptor);
            }
        }

        return results;
    }

    /**
     * Get dependency tree for a service
     */
    getDependencyTree(serviceName: string, visited = new Set<string>()): string[] {
        if (visited.has(serviceName)) {
            return []; // Circular dependency detected
        }

        visited.add(serviceName);
        const dependencies = this._dependencyGraph.get(serviceName) || new Set();
        const tree: string[] = [];

        for (const dependency of dependencies) {
            tree.push(dependency);
            tree.push(...this.getDependencyTree(dependency, new Set(visited)));
        }

        return tree;
    }

    /**
     * Detect circular dependencies
     */
    detectCircularDependencies(): string[][] {
        const cycles: string[][] = [];
        const visited = new Set<string>();
        const recursionStack = new Set<string>();

        const detectCycle = (serviceName: string, path: string[]): void => {
            if (recursionStack.has(serviceName)) {
                const cycleStart = path.indexOf(serviceName);
                cycles.push(path.slice(cycleStart).concat(serviceName));
                return;
            }

            if (visited.has(serviceName)) {
                return;
            }

            visited.add(serviceName);
            recursionStack.add(serviceName);

            const dependencies = this._dependencyGraph.get(serviceName) || new Set();
            for (const dependency of dependencies) {
                detectCycle(dependency, [...path, serviceName]);
            }

            recursionStack.delete(serviceName);
        };

        for (const serviceName of this._descriptors.keys()) {
            if (!visited.has(serviceName)) {
                detectCycle(serviceName, []);
            }
        }

        return cycles;
    }

    // ============================================================================
    // MAINTENANCE AND UTILITIES
    // ============================================================================

    /**
     * Clear all registrations
     */
    clear(): void {
        this._descriptors.clear();
        this._singletonInstances.clear();
        this._scopedInstances.clear();
        this._registrationOrder.length = 0;
        this._servicesByScope.clear();
        this._servicesByTag.clear();
        this._dependencyGraph.clear();
    }

    /**
     * Copy all registrations to another registry
     */
    copyTo(targetRegistry: ServiceRegistry): void {
        for (const [name, descriptor] of this._descriptors) {
            targetRegistry._descriptors.set(name, { ...descriptor });
        }

        for (const [name, instance] of this._singletonInstances) {
            targetRegistry._singletonInstances.set(name, instance);
        }

        targetRegistry._registrationOrder.push(...this._registrationOrder);

        // Copy indexes
        for (const [scope, services] of this._servicesByScope) {
            targetRegistry._servicesByScope.set(scope, new Set(services));
        }

        for (const [tag, services] of this._servicesByTag) {
            targetRegistry._servicesByTag.set(tag, new Set(services));
        }

        for (const [service, deps] of this._dependencyGraph) {
            targetRegistry._dependencyGraph.set(service, new Set(deps));
        }
    }

    /**
     * Get registry statistics
     */
    getStatistics() {
        const scopeCounts = new Map<ServiceScope, number>();
        for (const [scope, services] of this._servicesByScope) {
            scopeCounts.set(scope, services.size);
        }

        return {
            totalServices: this._descriptors.size,
            singletonInstances: this._singletonInstances.size,
            scopedInstances: this._scopedInstances.size,
            scopeCounts: Object.fromEntries(scopeCounts),
            registrationOrder: this._registrationOrder.length,
            circularDependencies: this.detectCircularDependencies().length
        };
    }

    // ============================================================================
    // PRIVATE UTILITIES
    // ============================================================================

    private _registerService<T>(
        serviceInterface: ServiceInterface<T>,
        implementation: Constructor<T>,
        scope: ServiceScope
    ): void {
        const descriptor: ServiceDescriptor<T> = {
            serviceInterface,
            implementation,
            scope,
            registeredAt: new Date(),
            metadata: serviceInterface.metadata
        };

        this._storeDescriptor(serviceInterface.name, descriptor);
        this._updateIndexes(serviceInterface.name, scope, serviceInterface.metadata);
        this._analyzeDependencies(serviceInterface.name, implementation);
    }

    private _storeDescriptor(serviceName: string, descriptor: ServiceDescriptor): void {
        this._descriptors.set(serviceName, descriptor);
        this._registrationOrder.push(serviceName);
    }

    private _updateIndexes(serviceName: string, scope: ServiceScope, metadata?: ServiceMetadata): void {
        // Update scope index
        let scopeSet = this._servicesByScope.get(scope);
        if (!scopeSet) {
            scopeSet = new Set();
            this._servicesByScope.set(scope, scopeSet);
        }
        scopeSet.add(serviceName);

        // Update tag indexes
        if (metadata?.tags) {
            for (const tag of metadata.tags) {
                let tagSet = this._servicesByTag.get(tag);
                if (!tagSet) {
                    tagSet = new Set();
                    this._servicesByTag.set(tag, tagSet);
                }
                tagSet.add(serviceName);
            }
        }
    }

    private _analyzeDependencies(serviceName: string, implementation: Constructor<any>): void {
        // Basic dependency analysis - in a real implementation, this would use
        // reflection or metadata to determine constructor dependencies
        const dependencies = new Set<string>();

        // For now, we'll extract dependencies from metadata if available
        const descriptor = this._descriptors.get(serviceName);
        if (descriptor?.metadata?.dependencies) {
            for (const dep of descriptor.metadata.dependencies) {
                dependencies.add(dep);
            }
        }

        this._dependencyGraph.set(serviceName, dependencies);
    }
}
